uploadNeuroDB/NeuroDB/MRI.pm

```diff
@@ -381,43 +354,93 @@ sub identify_scan_db {
     my $ScannerID = @$resultsRef> 0 ? $resultsRef->[0]->{'ID'} : 0;
     
-    # get the list of protocols for a site their scanner and subproject
+    #===========================================================#
+    # Get the list of lines in the mri_protocol table that      #
+    # apply to the given scan based on the center name          #
+    #===========================================================#
     $query = "SELECT *
               FROM mri_protocol
-              WHERE
-             (Center_name='$psc' AND ScannerID='$ScannerID')
-              OR ((Center_name='ZZZZ' OR Center_name='AAAA') AND ScannerID='0')
-              ORDER BY Center_name ASC, ScannerID DESC";
-
+              JOIN mri_protocol_group_target mpgt USING (MriProtocolGroupID)
+              WHERE (
+                        (Center_name = ? AND ScannerID = ?)
+                     OR ((Center_name='ZZZZ' OR Center_name='AAAA') AND ScannerID='0'))";
+
+    #============================================================#
+    # Add to the query the clause related to the Project ID, the #
+    # subproject ID and visit label.                             #
+    #============================================================#
+    $query .= defined $projectID
+        ? ' AND (mpgt.ProjectID IS NULL OR mpgt.ProjectID = ?)'
+        : ' AND mpgt.ProjectID IS NULL';
+    $query .= defined $subprojectID
+        ? ' AND (mpgt.SubprojectID IS NULL OR mpgt.SubprojectID = ?)'
+        : ' AND mpgt.SubprojectID IS NULL';
+    $query .= defined $visitLabel
+        ? ' AND (mpgt.Visit_label IS NULL OR mpgt.Visit_label = ?)'
+        : ' AND mpgt.Visit_label IS NULL';
+        
+    $query .=  ' ORDER BY Center_name ASC, ScannerID DESC';
+
+    my @bindValues = ($psc, $ScannerID);
+    push(@bindValues, $projectID)    if defined $projectID;
+    push(@bindValues, $subprojectID) if defined $subprojectID;
+    push(@bindValues, $visitLabel)   if defined $visitLabel;
+    
     $sth = $${dbhr}->prepare($query);
-    $sth->execute();
-    return 'unknown' unless $sth->rows>0;
+    $sth->execute(@bindValues);
+    
+    my @rows = @{ $sth->fetchall_arrayref({}) };
+    # If no lines in the mri_protocol_group_target matches the ProjectID/SubprojectID/VisitLabel
+    # then no lines of the mri_protocol table can be used to identify the scan type. This is most
+    # likely a setup issue: mri_protocol/mri_protocol_group/mri_protocol_group_target do not cover
+    # all the cases. Warn.
+    if(@rows == 0) {
+		my $msg = "Warning! No protocol group can be used to determine the scan type "
+		        . "of $minc_location.\nIncorrect/incomplete setup of table mri_protocol_group_target: "
+		        . " setting scan type to 'unknown'"; 
+		print "$msg\n";
+        my $notify = NeuroDB::Notify->new( $dbhr );
+        $notify->spool('mri upload processing class', $msg, 0, 'MRI.pm', $uploadID, 'N', 'Y');
+
+        return 'unknown' unless $sth->rows>0;
```

Comment:
alignment is off

Reply:
not sure I understand why there is an `unless $sth->rows>0` since we already established here that `@rows == 0`

Reply:
Fixed alignment. The `unless` is useless: removed.